pkg/documentlog/impl.go

```diff
@@ -108,158 +86,132 @@ func (dl *documentLog) Subscribe(documentType string) DocumentQueue {
 }
 
-func (dl *documentLog) AddDocumentWithContents(timestamp time.Time, documentType string, contents io.Reader) (model.Document, error) {
+func (dl *documentLog) AddDocumentWithContents(timestamp time.Time, documentType string, contents io.Reader) (*model.Document, error) {
 	buffer := new(bytes.Buffer)
 	if _, err := buffer.ReadFrom(contents); err != nil {
-		return model.Document{}, err
+		return nil, err
 	}
 	document := model.Document{
 		Hash:      model.CalculateDocumentHash(documentType, timestamp, buffer.Bytes()),
 		Type:      documentType,
 		Timestamp: timestamp,
 	}
-	if dl.HasContentsForDocument(document.Hash) {
-		return model.Document{}, fmt.Errorf("document already exists (with content) for hash: %s", document.Hash)
+	if err := dl.addDocument(document); err != nil {
+		return nil, err
 	}
-	dl.AddDocument(document)
 	return dl.AddDocumentContents(document.Hash, buffer)
 }
 
-func (dl *documentLog) AddDocument(document model.Document) {
-	dl.documentsMutex.WriteLock(func() {
-		existing := dl.documentHashIndex[document.Hash.String()]
-		if existing != nil {
-			// Hash already present, but check if the timestamp matches, just to be sure
-			t := existing.Document.Timestamp.UnixNano()
-			if t != document.Timestamp.UnixNano() {
-				log.Log().Warnf("Integrity violation! Document hash %s with timestamp %d is already present with different timestamp (%d)", document.Hash, document.Timestamp.UnixNano(), t)
-			}
-			return
-		}
-		newEntry := &entry{Document: document}
-		dl.entries = append(dl.entries, newEntry)
-		dl.documentHashIndex[newEntry.Hash.String()] = newEntry
-		// TODO: Isn't there a faster way to keep it sorted (or not sort it at all?)
-		sort.Slice(dl.entries, func(i, j int) bool {
-			if dl.entries[i].Timestamp == dl.entries[j].Timestamp {
-				return dl.entries[i].Timestamp.Before(dl.entries[j].Timestamp)
-			} else {
-				return dl.entries[i].Hash.String() < dl.entries[j].Hash.String()
-			}
-		})
-		// Calc last consistency hash
-		// TODO: Test this
-		// TODO: Make this smarter (retain unchanged consistency hashes)
-		dl.consistencyHashIndex = make(map[string]*entry, len(dl.entries))
-		documents := make([]model.Document, len(dl.entries))
-		var i = 0
-		prevHash := model.EmptyHash()
-		consistencyHashes := make([]string, len(dl.entries))
-		for i = 0; i < len(dl.entries); i++ {
-			documents[i] = dl.entries[i].Document
-			if i == 0 {
-				copy(prevHash, dl.entries[i].Hash)
-			} else {
-				prevHash = model.MakeConsistencyHash(prevHash, dl.entries[i].Hash)
-			}
-			consistencyHashes[i] = prevHash.String()
-			dl.consistencyHashIndex[prevHash.String()] = dl.entries[i]
+func (dl *documentLog) addDocument(document model.Document) error {
+	existing, err := dl.store.Get(document.Hash)
+	if err != nil {
+		return err
+	}
+	if existing != nil {
+		// Hash already present, but check if the timestamp matches, just to be sure
+		t := existing.Document.Timestamp.UnixNano()
+		if t != document.Timestamp.UnixNano() {
```

Comment:
This should actually never happen since it means our database (or the record we received) is corrupted, since a different timestamp should lead to a different hash. So this is a if-all-else-fails guard.